src/runner.py

```diff
@@ -105,11 +101,21 @@ def _cleanup_paths(self):
         self._job_tmp_dirs = job_tmp_dirs
         # ToDo: if stat != 0 then report error to API?
 
-    def loop(self):
+    def _setup(self, args):
+        ctx = Service.Context()
         sub_id = self._db.subscribe_node_channel(filters={
             'name': 'checkout',
             'state': 'available',
         })
+        return Service.Context(sub_id=sub_id)
+
+    def _stop(self, ctx):
+        sub_id = getattr(ctx, 'sub_id', None)
+        if sub_id:
```

Comment:
I think that with current implementation it's guaranteed that `sub_id` is present and have a default value of None

```suggestion
        if ctx.sub_id:
```